File: 675/solution.go
package main /*{{{*/

import (
	"fmt"
	"reflect"
)

const (
	e4  = int64(10000)
	e5  = int64(100000)
	e6  = int64(1000000)
	e7  = int64(10000000)
	e8  = e4 * e4
	mod = 1e9 + 87
)

func add(list ...interface{}) int64 { /*{{{*/
	if len(list) == 0 {
		return 0
	}
	ret := int64(0)
	for _, i := range list {
		buf := int64(0)
		switch reflect.TypeOf(i).String() {
		case "int":
			buf = int64(i.(int))
		case "int64":
			buf = i.(int64)
		}
		ret = (ret + buf) % mod
	}
	return ret
} /*}}}*/
func mul(list ...interface{}) int64 { /*{{{*/
	if len(list) == 0 {
		return 0
	}
	ret := int64(1)
	for _, i := range list {
		buf := int64(0)
		switch reflect.TypeOf(i).String() {
		case "int":
			buf = int64(i.(int))
		case "int64":
			buf = i.(int64)
		}
		ret = (buf * ret) % mod
	}
	return ret
} /*}}}*/
/*}}}*/

var p = [e7 + 10]bool{}
var l = [e7 + 10]int64{}
var cnt = [e7 + 10]int64{}
var inv = [4*e7 + 10]int64{}

func init() { /*{{{*/
	inv[1] = 1
	for i := int64(2); i <= 4*e7; i++ {
		inv[i] = mul(add(mod, mod-mod/i), inv[mod%i])
	}
	for i := int64(2); i <= e7; i++ {
		if !p[i] {
			for j := i; j <= e7; j += i {
				l[j] = i
				p[j] = true
			}
		}
	}
} /*}}}*/

var base = int64(1)

func cal(a int64) int64 {
	for a != 1 {
		x := add(1, mul(2, cnt[l[a]]))
		base = mul(base, inv[x])
		base = mul(base, add(1, mul(2, cnt[l[a]]), 2))
		cnt[l[a]]++
		a /= l[a]
	}
	return base
}
func main() {
	ans := int64(0)
	for i := int64(2); i <= e7; i++ {
		ans = add(ans, cal(i))
	}
	fmt.Println(ans)
}

File: 675/solution.cpp
#include "../template/bigint.cpp"
#include <iostream>/*{{{*/
#include <map>
#include <math.h>
#include <set>
#include <vector>
const int e2  = 100;
const int e3  = e2 * 10;
const int e4  = e3 * 10;
const int e5  = e4 * 10;
const int e6  = e4 * e2;
const int e7  = e6 * 10;
const int e8  = e7 * 10;
const int mod = 1e9 + 87;
using namespace std; /*}}}*/

/*add{{{*/
template <typename T> T base_add(T a, T b) {
    a %= mod;
    b %= mod;
    return (a + b) % mod;
}
template <typename T> T add(const T &t) { return t; }
template <typename T, typename... Args> T add(const T &t, const Args &...args) { return base_add(t, add(args...)); }
/*}}}*/

/*{{{mul*/
template <typename T> T base_mul(T a, T b) {
    a %= mod;
    b %= mod;
    T ret(0);
    while (b) {
        if (b & 1)
            ret = (ret + a) % mod;
        b >>= 1;
        a = (a + a) % mod;
    }
    return ret;
}
template <typename T> T mul(const T &t) { return t; }
template <typename T, typename... Args> T mul(const T &t, const Args &...args) { return base_mul(t, mul(args...)); }
/*}}}*/

/*{{{qpow*/
template <typename T> T qpow(T a, T b) {
    T ret(1);
    while (b) {
        if (b & 1)
            ret = mul(ret, a);
        a = mul(a, a);
        b >>= 1;
    }
    return ret;
}
/*}}}*/

bool p[e7 + 10];
int l[e7 + 10]; // 对i来说最小的prime
int cnt[e7 + 10];
int inv[4 * e7 + 10];
void init() {
    inv[1]    = 1;
    int e7__2 = e7 << 2;
    for (int i = 2; i <= e7__2; i++) {
        inv[i] = mul(mod - mod / i, inv[mod % i]);
    }
    for (int i = 2; i <= e7; i++)
        if (!p[i]) {
            for (int j = i; j <= e7; j += i) {
                l[j] = i;
                p[j] = true;
            }
        }
    cout << "init success" << endl;
}
int base = 1;
int cal(int a) {
    while (a != 1) {
        int x = add(1, mul(2, cnt[l[a]]));
        base  = mul(base, inv[x]);
        base  = mul(base, add(1, mul(2, cnt[l[a]]), 2));
        cnt[l[a]]++;
        a /= l[a];
    }
    return base;
}
int main() {
    init();
    int ans = 0;
    for (int i = 2; i <= e7; i++)
        ans = add(ans, cal(i));
    cout << ans << endl;
    return 0;
}
